### ISSUE TITLE
Securely pass credentials to a Containerized KRM

### ISSUE THREAD

**Issue by mdbooth** [kubernetes-sigs/kustomize#5152]
### Eschewed features

- [X] This issue is not requesting templating, unstuctured edits, build-time side-effects from args or env vars, or any other eschewed feature.

### What would you like to have added?

Some mechanism to pass arbitrary authentication data to a a Containerized KRM which doesn't require storing them in the kustomize directory.

This is likely to have to include environment variables and user config files. While this is technically _build-time side-effects from args or env vars_, the purpose is to provide static content. The effect is identical to including the output on disk with the advantage that it is not on disk, but the disadvantage is that you may need additional sensitive information to obtain it.

This is explained better below.

### Why is this needed?

I would like to create a KRM function for generating Secrets which obtains the secret payload from an external secret management system. The purpose of this is to avoid storing deployment secrets in a file or SCM along with other kustomize data, while still allowing a complete deployment using kustomize. My personal use case is manual deployment from the command line, although there may also be cases where this makes sense in automated tooling.

My kustomize will contain something like:
```yaml
apiVersion: openstack-barbican.example.com/v1alpha1
kind: SecretGenerator
metadata:
  name: notImportantHere
  annotations:
    config.kubernetes.io/function: |
      container:
        image: quay.io/mdbooth/kustomize-barbican:latest
spec:
  name: passwords
  values:
  - name: root
    ref: https://x.x.x.x:13311/v1/secrets/aaaa-bbbb-cccc-dddd
```

It will generate:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: passwords
data:
  root: c3VwZXJzZWNyZXQ=
type: Opaque
```

### Can you accomplish the motivating task without this feature, and if so, how?

Yes. I can use an Exec KRM. However, my strong preference is to write a containerized KRM as it's simpler to distribute, and is also the explicit recommendation in https://kubectl.docs.kubernetes.io/guides/extending_kustomize/exec_krm_functions/. An exec function must be embedded within the Kustomization's directory, which is not ideal for a generic tool which can be used anywhere a particular secret management system is deployed.

However, obtaining data from a secret management system inevitably requires data from the environment to provide, e.g. endpoints and authentication. These can be in the form of either environment variables or files. It's possible they could also be access to a hardware token, which could be very difficult to support. As far as I can tell there is no way to pass this type of information to a Containerized KRM without putting it in a file, which would defeat the primary purpose of the tool.

### What other solutions have you considered?

As above, an Exec KRM. I could also deploy secrets to my environment using a tool which is independent of kustomize.

If we decided that this was best implemented in an Exec KRM (e.g. so we don't even need to consider hardware token access), my request would be for better options for distributing an Exec KRM.

### Anything else we should know?

I have not proposed any specific solution because I am not familiar enough with the design space.

Reading 'Eschewed Features' it is clear that the project wants deterministic output. My use case is consistent with this goal, but providing the functionality required to implement my use case may enable use cases which the project doesn't appear to want to support, e.g. setting a VERSION in an object from a build-time environment variable. This is already possible with an Exec KRM, though, so perhaps it could be viewed in that light.

While I don't feel competent to design an API change in a new project, I may be able to find the time to implement one designed by others.

### Feature ownership

- [X] I am interested in contributing this feature myself! 🎉

**Comment by annasong20** [comment]
Hi @mdbooth, thank you for your detailed write-up. Unfortunately, because KRM functions are an alpha feature, thus subject to change, they're lower priority and we're hesitant to violate kustomize's fundamental "no build-time side effects" principle to change it. Please see the following discussions on kustomize's stance on "build-time side effects":
* https://github.com/kubernetes-sigs/kustomize/issues/4731#issuecomment-1423123566
* https://github.com/kubernetes-sigs/kustomize/issues/4655#issuecomment-1155605705

Additionally, we hope you find the alternatives in this issue helpful to your authentication use case: #4633. Please re-open if you'd like to further discuss.

/triage out-of-scope
/close

**Comment by mdbooth** [comment]
I was expecting this response, tbh. I don't agree with your argument in this case: as explained in the description the feature is not intended to introduce build time side effects. Well failure is technically a side effect, but if it produces any output at all the result is deterministic. It is as deterministic as the content of a file, except with the benefit of using whatever secure storage solution you have for secret management.

Additionally I think alpha is the perfect time to experiment with this kind of functionality as it can still be relatively safely removed if the costs turn out to outweigh the benefits in practise.

However, as I don't have time to work on a feature I would also have to advocate for I'm not going to press the matter.

**Comment by annasong20** [comment]
@mdbooth thank you for the explanation. My understanding is that kustomize has strong opinions regarding any type of `build-time side effects`, which seem inevitable in your proposition. 

However, this is also my first time personally encountering an issue related to this principle. I'll take some more time to think about how your use case lines up with kustomize's principles. Sorry and thank you for your patience.

**Comment by annasong20** [comment]
Hi @mdbooth, after thinking more about this use case, I really empathize with your argument that

> Well failure is technically a side effect, but if it produces any output at all the result is deterministic.

I believe we're **potentially open** to accepting "build-time side effects" **if** the operations that kustomize performs are deterministic, but the "build-time side effects", such as local environment variables, are only used to determine whether or not the operations succeed. A concrete example of this is using environment variables to authenticate to private repositories. If the authentication fails, `kustomize build` should return an error. Just as an aside, I'm thinking to guarantee that environmental variables are only used for such purposes, we may have to add fields to the kustomization file dedicated to specifying and reading from such variables.

However, please do not treat this as a **guarantee** that we will accept such "build-time side effects" proposals. As stated previously, repeatedly, kustomize feels very strongly about protecting the tool from "build-time side effects" and including them even for a very specific purpose is a major decision. We have to consider its repercussions and unintended side effects on not only kustomize itself, but also on upstream [kubectl](https://github.com/kubernetes/kubectl). As such, I believe whoever has a more concrete design for such a feature should propose it in the form of a full-fledged KEP ([Kubernetes Enhancement Proposal](https://github.com/kubernetes/enhancements)). A mini kustomize KEP: https://github.com/kubernetes-sigs/kustomize/tree/master/proposals will not suffice.

For reference, @natasha41575

**Comment by mdbooth** [comment]
Shal I reopen?

**Comment by annasong20** [comment]
/reopen

@mdbooth Sorry I forgot!

**Comment by btrepp** [comment]
I am still very much still struggling with secrets and kustomize as well.
The core principles of kustomize I do agree with, but it's an incredibly strange experience with any sensitive data. Where foot guns abound plenty.

I think your suggestion there with http may be the path forward (or something like it). It also shows an inconsistency in kustomize. The deterministic rules only really apply to local files, HTTPs/git based items don't have that restriction, they are allowed to be undeterministic. So you could probably abuse that to have secrets outside of the kustomize directory, but with the load that you need to run a HTTP server.

I got close with having ninja generate kustomize patches (put also needed to write a custom tool to discover kustomize dependencies, so they could then be read), however I was storing them out-of-tree (so you know, I don't accidentally commit them to source, but then I hit this https://github.com/kubernetes-sigs/kustomize/issues/865

Funnily enough, working around this using say a local http server, makes kustomize more undeterministic. Ninja knew when the secrets files changed, so could use that, but kustomize doesn't like that as password store is 'out of the build directory'. I'm also not a fan of throwing the baby out with the bathwater with that flag.

I'm beginning to think if kustomize had the ability to read from sockets (like say HTTP over a socket, instead of TCP) that might go a long way. There would be security, the user would have be be authorised to read the socket anyway, so better than a local http server. There's no loss in determinism either, this is just as deterministic as HTTP/git ones, as they offer zero guarantees as to whether the output is the same currently, and there's no 'secrets in git' problem/foot-guns awaiting a accidentally `git add .` to compromise your apps security.

I think this could also work well with just being 'patches'. E.g you create an empty secret with secret generator, but use the socket to retrieve a patch that kustomize would apply.

To even restrict it further, these could possibly only be allowed in $HOME/.kustomize/xyz.sock, or something, which locks down the ability to read other than what the user specified. I am not sure if that's necessary though

**Comment by mdbooth** [comment]
Somebody mentioned this project to me recently: https://external-secrets.io/

It's still on my todo list to investigate properly, but my initial understanding is that you'd store an ExternalSecret CR in git, and the operator would hydrate it at 'runtime'. I think this implies that the secret store credentials are present in the cluster, though, and I'm not currently sure what I think about that. It would be a very tidy way to do this with kustomize, though.

**Comment by btrepp** [comment]
@mdbooth I have tried that myself, and moved away. It is a very cool concept though!.

One of the issues with external-secrets in kustomize, is you loose the *brilliant* hashing functionality on the secrets. Which has the side-effect that the deployment gets updated and re-rolled out. I ended up a rabbit hole of trying to have some extra controller that reboot containers on secret change.

An option I have switched to over the fifos (couldn't get that to play nice with ninja).

This abuses the fact that kustomize can go 'out of tree' for components, and suck them back in.

live/app/kustomization.yaml
```

resources: .....
components:
- ../../out/components/system-secrets
```

out/components/system-secrets/kustomization.yaml
```
apiVersion: kustomize.config.k8s.io/v1alpha1
kind: Component

patches:
- path: authentik.yaml
```

Where there is a patch file there.

This way an 'external tool', can completely generate those patches/resources, then have Kustomize pick up on them, this retains the 'out of tree' behaviour that is important for secrets.

I've folded that in with my ninja build setup, and it works quite well!. So I've finally got something that works for me after a year of banging my head on this. The setup is here https://git.sr.ht/~btrepp/environment/tree, though admittedly maybe not the easiest to follow at the moment
